qt/widgets/common/src/WorkspaceMultiSelector.cpp

```diff
@@ -0,0 +1,433 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+//------------------------------------------------------
+// Includes
+//------------------------------------------------------
+#include "MantidQtWidgets/Common/WorkspaceMultiSelector.h"
+
+#include <Poco/AutoPtr.h>
+#include <Poco/NObserver.h>
+#include <Poco/Notification.h>
+#include <Poco/NotificationCenter.h>
+
+#include "MantidAPI/AlgorithmManager.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/WorkspaceGroup.h"
+
+#include <QCompleter>
+#include <QDebug>
+#include <QHeaderView>
+#include <QLineEdit>
+#include <QStyledItemDelegate>
+
+constexpr short namesCol = 0;
+constexpr short indexCol = 1;
+
+namespace {
+using namespace Mantid::API;
+
+QStringList headerLabels = {"Workspace Name", "Ws Index"};
+
+MatrixWorkspace_sptr getWorkspace(const std::string &name) {
+  return AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(name);
+}
+
+boost::optional<std::size_t> maximumIndex(const MatrixWorkspace_sptr &workspace) {
+  if (workspace) {
+    const auto numberOfHistograms = workspace->getNumberHistograms();
+    if (numberOfHistograms > 0)
+      return numberOfHistograms - 1;
+  }
+  return boost::none;
+}
+
+QString getIndexString(const MatrixWorkspace_sptr &workspace) {
+  const auto maximum = maximumIndex(workspace);
+  if (maximum) {
+    if (*maximum > 0)
+      return QString("0-%1").arg(*maximum);
+    return "0";
+  }
+  return "";
+}
+
+QString getIndexString(const std::string &workspaceName) { return getIndexString(getWorkspace(workspaceName)); }
+
+std::unique_ptr<QRegExpValidator> createValidator(const QString &regex, QObject *parent) {
+  return std::make_unique<QRegExpValidator>(QRegExp(regex), parent);
+}
+
+QString OR(const QString &lhs, const QString &rhs) { return "(" + lhs + "|" + rhs + ")"; }
+
+QString NATURAL_NUMBER(std::size_t digits) { return OR("0", "[1-9][0-9]{," + QString::number(digits - 1) + "}"); }
```

Comment:
A lot of these functions/variables in the annonymous namespace are also in the `AddWorkspaceDialog` file. Is there a way we can avoid duplicating this code?

Reply:
I have seen this exact regexes populating many anonymous namespace, Would you agree to open this as a separate issue? I would be glad to work on that for next sprint.

Reply:
Yes I'm happy for this to be a separate issue